pkg/controller/utils/utils_test.go

```diff
@@ -111,3 +123,213 @@ func TestLogLevel(t *testing.T) {
 		assert.Equal(t, tc.expectedLevel, actualLevel)
 	}
 }
+
+func TestGetConcurrentReconciles(t *testing.T) {
+	cases := []struct {
+		name                         string
+		environmentVariables         map[string]string
+		expectedConcurrentReconciles int
+		expectedError                bool
+	}{
+		{
+			name: "Only default goroutines is set",
+			environmentVariables: map[string]string{
+				fmt.Sprintf(ConcurrentReconcilesEnvVariableFormat, "default"): "10",
+			},
+			expectedConcurrentReconciles: 10,
+		},
+		{
+			name: "Only controller goroutines is set",
+			environmentVariables: map[string]string{
+				fmt.Sprintf(ConcurrentReconcilesEnvVariableFormat, testControllerName): "10",
+			},
+			expectedConcurrentReconciles: 10,
+		},
+		{
+			name: "Both default and controller goroutines are set",
+			environmentVariables: map[string]string{
+				fmt.Sprintf(ConcurrentReconcilesEnvVariableFormat, "default"):          "10",
+				fmt.Sprintf(ConcurrentReconcilesEnvVariableFormat, testControllerName): "20",
+			},
+			expectedConcurrentReconciles: 20,
+		},
+		{
+			name:                         "Neither default nor controller goroutines are set",
+			environmentVariables:         map[string]string{},
+			expectedConcurrentReconciles: concurrentControllerReconciles,
+		},
+		{
+			name: "default goroutines are set incorrectly",
+			environmentVariables: map[string]string{
+				fmt.Sprintf(ConcurrentReconcilesEnvVariableFormat, "default"): "not-a-int",
+			},
+			expectedError: true,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// set environment variables
+			for k, v := range tc.environmentVariables {
+				os.Setenv(k, v)
+			}
+
+			concurrentReconciles, err := getConcurrentReconciles(testControllerName)
+			if tc.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+			assert.Equalf(t, tc.expectedConcurrentReconciles, concurrentReconciles, "unexpected concurrent reconciles")
+			// unset environment variables
+			for k := range tc.environmentVariables {
+				os.Unsetenv(k)
+			}
```

Comment:
This should be in a defer after setting the environment variables needed for the test case.